File: code/packages/elixir/compiler_ir/lib/coding_adventures/compiler_ir/printer.ex
defmodule CodingAdventures.CompilerIr.Printer do
  @moduledoc """
  IR Printer — `IrProgram` → human-readable text.

  The printer converts an `IrProgram` into its canonical text format.
  This format serves three purposes:

  1. **Debugging** — humans can read the IR to understand what the
     compiler produced.
  2. **Golden-file tests** — expected IR output is committed as `.ir`
     text files and compared against actual output.
  3. **Roundtrip** — `parse(print(program)) == program` is a testable
     invariant.

  ## Text format

      .version 1

      .data tape 30000 0

      .entry _start

      _start:
        LOAD_ADDR   v0, tape          ; #0
        LOAD_IMM    v1, 0             ; #1
        HALT                          ; #2

  ## Key rules

  - `.version N` is always the first non-comment line.
  - `.data` declarations come before `.entry`.
  - Labels are on their own unindented line with a trailing colon.
  - Instructions are indented with two spaces.
  - `; #N` comments show instruction IDs (informational only).
  - `COMMENT` instructions emit as `; <text>` (no ID).
  """

  alias CodingAdventures.CompilerIr.{IrProgram, IrInstruction, IrDataDecl, IrOp}
  alias CodingAdventures.CompilerIr.{IrRegister, IrImmediate, IrLabel}

  @doc """
  Convert an `IrProgram` to its canonical text representation.

  ## Examples

      iex> p = IrProgram.new("_start")
      iex> text = Printer.print(p)
      iex> String.starts_with?(text, ".version 1")
      true
  """
  @spec print(IrProgram.t()) :: String.t()
  def print(%IrProgram{} = program) do
    parts = [
      # Version directive — always first
      ".version #{program.version}\n",

      # Data declarations — one per .data line
      print_data(program.data),

      # Entry point directive
      "\n.entry #{program.entry_label}\n",

      # Instructions
      print_instructions(program.instructions)
    ]

    Enum.join(parts)
  end

  # ── Private helpers ──────────────────────────────────────────────────────────

  # Emit each .data declaration on its own line, preceded by a blank line.
  defp print_data([]), do: ""

  defp print_data(decls) do
    Enum.map_join(decls, "", fn %IrDataDecl{label: lbl, size: sz, init: init} ->
      "\n.data #{lbl} #{sz} #{init}\n"
    end)
  end

  # Emit all instructions.
  defp print_instructions(instructions) do
    Enum.map_join(instructions, "", &print_instruction/1)
  end

  # Label instructions get their own unindented line with a trailing colon.
  # They produce no machine code, so no ID comment.
  defp print_instruction(%IrInstruction{opcode: :label, operands: [operand]}) do
    "\n#{operand_to_string(operand)}:\n"
  end

  # COMMENT instructions emit as "; <text>" — no ID comment.
  defp print_instruction(%IrInstruction{opcode: :comment, operands: operands}) do
    text =
      case operands do
        [op | _] -> operand_to_string(op)
        [] -> ""
      end

    "  ; #{text}\n"
  end

  # Regular instructions: "  OPCODE  op1, op2, ...  ; #ID"
  defp print_instruction(%IrInstruction{opcode: opcode, operands: operands, id: id}) do
    op_name = IrOp.to_string(opcode)

    # Pad the opcode name to 11 characters for visual alignment.
    # E.g. "LOAD_ADDR  " (11 chars), "HALT       " (11 chars)
    padded_name = String.pad_trailing(op_name, 11)

    operand_str =
      operands
      |> Enum.map(&operand_to_string/1)
      |> Enum.join(", ")

    "  #{padded_name}#{operand_str}  ; ##{id}\n"
  end

  # Convert an operand struct to its text representation.
  defp operand_to_string(%IrRegister{} = r), do: IrRegister.to_string(r)
  defp operand_to_string(%IrImmediate{} = i), do: IrImmediate.to_string(i)
  defp operand_to_string(%IrLabel{} = l), do: IrLabel.to_string(l)
end

File: code/packages/ruby/compiler-ir/lib/coding_adventures/compiler_ir/printer.rb
# frozen_string_literal: true

# ==========================================================================
# IrPrinter — IrProgram → Human-Readable Text
# ==========================================================================
#
# The printer converts an IrProgram into its canonical text format.
# This format serves three purposes:
#
#   1. Debugging — humans can read the IR to understand what the compiler did
#   2. Golden-file tests — expected IR output is committed as .ir text files
#   3. Roundtrip — parse(print(program)) == program is a testable invariant
#
# ── Text Format ─────────────────────────────────────────────────────────────
#
#   .version 1
#
#   .data tape 30000 0
#
#   .entry _start
#
#   _start:
#     LOAD_ADDR   v0, tape          ; #0
#     LOAD_IMM    v1, 0             ; #1
#     HALT                          ; #2
#
# Key rules:
#   - .version N is always the first non-comment line
#   - .data declarations come before .entry
#   - Labels are on their own unindented line with a trailing colon
#   - Instructions are indented with two spaces
#   - ; #N comments show instruction IDs (informational only)
#   - COMMENT instructions emit as "; <text>" on their own indented line
#
# The opcode field is left-padded to 11 characters so that operands align
# into a readable column. This matches the Go printer's "%-11s" format.
# ==========================================================================

module CodingAdventures
  module CompilerIr
    module IrPrinter
      # print(program) → String
      #
      # Converts an IrProgram to its canonical text representation.
      # The returned string contains the full .ir file content including
      # the version directive, data declarations, entry point, and all
      # instructions with their ID comments.
      #
      # @param program [IrProgram] the program to print
      # @return [String] the canonical IR text
      def self.print(program)
        parts = []

        # Version directive — always the first line.
        parts << ".version #{program.version}"

        # Data declarations — one per line, blank line before each.
        program.data.each do |decl|
          parts << ""
          parts << ".data #{decl.label} #{decl.size} #{decl.init}"
        end

        # Entry point — blank line before.
        parts << ""
        parts << ".entry #{program.entry_label}"

        # Instructions — each on its own line.
        program.instructions.each do |instr|
          case instr.opcode
          when IrOp::LABEL
            # Labels get their own unindented line with a trailing colon.
            # A blank line before each label visually separates basic blocks.
            parts << ""
            parts << "#{instr.operands[0]}:"

          when IrOp::COMMENT
            # COMMENT instructions emit as "  ; <text>" — indented like
            # regular instructions but without an opcode mnemonic.
            text = instr.operands[0]&.to_s || ""
            parts << "  ; #{text}"

          else
            # Regular instruction: "  OPCODE       operands  ; #ID"
            #
            # The opcode mnemonic is left-aligned in an 11-character field
            # so that operands form a readable column. The %-11s format in
            # Ruby's sprintf matches Go's fmt.Sprintf("%-11s", ...).
            mnemonic = IrOp.op_name(instr.opcode)
            operand_str = instr.operands.map(&:to_s).join(", ")
            line = format("  %-11s", mnemonic)
            line += operand_str unless operand_str.empty?
            line += "  ; ##{instr.id}"
            parts << line
          end
        end

        # Join all parts with newlines and add a trailing newline.
        parts.join("\n") + "\n"
      end
    end
  end
end
